### DISCUSSION TITLE
How can I make an action run if build is successful or unneeded

### DISCUSSION THREAD

**Discussion by mark-summerfield** [SCons/scons#4582] [Q&A] [answered]
I want to make an action run if a build is successful _or_ if the build isn't needed (i.e., no source changes).

This will make an action run if the build is successful:

```
AddOption('--run', dest='run', action='store_true', default=False,
          help='Run once built')
# etc ...
if GetOption('run'):
    env.AddPostAction(app, Action('./$TARGET'))
```
This allows me to enter `scons --run` to build & run. But if I up-arrow in the terminal to rerun this command, scon says (rightly) that there's nothing to build and does _not_ do the post option.

How can I make it do the post option in this second case?

**Comment by mwichmann** [comment]
You've already realized that a post (or pre) action attached to a target doesn't trigger if the target isn't rebuilt. From your description, it's not clear why you want to run something even if the target wasn't built... if you mean you want to actually execute the command (like *might* be the case for unit tests), you have to take a bit of a different approach.  Don't take this as the definitive answer, but you could try this (there are several ways to skin this cat, and this may not be the best):

1. construct a `Command` builder call which runs the program you're interested in. 
2. mark the target of the `Command` as depending on the original generated program.
3. add the target to an `Alias` called `run`
4. mark the alias as `AlwaysBuild` so as long that (alias) target is selected, it will build even if it doesn't seem "out of date"

Now build as `scons run`

**Reply by mark-summerfield** [reply]
I understand the theory of what you're saying, but I can't see how to do it in code (I'm an scons beginner).

**Comment by bdbaddog** [comment]
You could just setup an atexit.

See: https://docs.python.org/3/library/atexit.html for info
Basically you just set up a function to run your program, and then register it with `atexit.register`

@mwichmann's is another scons-only way to do this.

**Reply by mark-summerfield** [reply]
atexit won't help because I want to run if there is a successful build or if a build isn't needed, but _not_ if a build fails.
This will allow me to write 'scons run' and then just up-arrow and Enter after each change I want to test.

**Comment by mwichmann** [comment, marked as answer]
At its very simplest, this should work:
```py
prog = Program("hello.c")
pname = str(prog[0])
runit = Command(target=f'r_{pname}', source=prog, action=(f'-./{pname}'))
```
In this case, `Command` has a target that uses a made-up name, just prepending something to the name of the `Program` target, and its source is the built program, and since the action never "creates" that target, it's always out of date and so runs every time.

The more detailed recipe accounts for more complexity, such as when there are multiple programs to run (the accumulating nature of Alias is useful for that), or when the action actually produces something that looks like a target - like an output log (if you use a logfile or other tangible thing as the runner target, that's where `AlwaysBuild` would come into play). The `Depends` call from the recipe isn't needed if you are okay having the built program as the source, otherwise you'd need it.  

SCons by default selects all targets under the project root, so in the minimal example both building the program and running it are selected. If you want the default to be "build" and "running" to have to be explicitly asked for, you can expand the recipe a bit:

```py
prog = Program("hello.c")
Default(prog)
pname = str(prog[0])
runit = Command(target=f'r_{pname}', source=prog, action=(f'-./{pname}'))
alias = Alias("run", runit)
```
which might look like:
```
$ scons -Q
gcc -o hello.o -c hello.c
gcc -o hello hello.o
$ scons -Q
scons: `hello' is up to date.
$ scons -Q run
./hello
Hello world
$ rm hello
$ scons -Q run
gcc -o hello hello.o
./hello
Hello world
```

I'd recommend taking the simplest version and running some experiments, like build with `--tree=all` to make sure the plumbing looks right, and don't get more complicated unless you need to.

**Reply by mark-summerfield** [reply]
Based on your ideas I'm using the below (which only needs one line to be changed per project):
```
AddOption('--dbg', dest='dbg', action='store_true', default=False,
          help='Debug build')
env = Environment()
if GetOption('dbg'):
    env.Append(CCFLAGS='-Wall -g')
else:
    env.Append(CCFLAGS='-Wall -s -O3 -DNDEBUG')
env.Append(CPPPATH=['.', '../STC/include'])

app = env.Program('myapp', ['myapp.c']) # only this changes

Default(app)
appname = str(app[0])
runner = Command(target=f'run_{appname}', source=app,
                 action=f'-./{appname}')
alias = Alias('run', runner)
```
Now I can run `scons -Q run` to build & run, and to run.

Thanks.
